kernel/comps/block/src/lib.rs

```diff
@@ -75,51 +98,69 @@ impl dyn BlockDevice {
     }
 }
 
-pub fn register_device(name: String, device: Arc<dyn BlockDevice>) {
-    COMPONENT
-        .get()
-        .unwrap()
-        .block_device_table
-        .lock()
-        .insert(name, device);
+/// The error type which is returned from the APIs of this crate.
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub enum Error {
+    /// Already registered
+    Registered,
+    /// Not found
+    NotFound,
+    /// Invalid arguments
+    InvalidArgs,
+    /// Id Exhausted
+    IdExhausted,
 }
 
-pub fn get_device(str: &str) -> Option<Arc<dyn BlockDevice>> {
-    COMPONENT
-        .get()
-        .unwrap()
-        .block_device_table
+/// Registers a new block device.
+pub fn register(device: Arc<dyn BlockDevice>) -> Result<(), Error> {
+    let mut registry = DEVICE_REGISTRY.lock();
+    let id = device.id().as_encoded_u64();
+    if registry.contains_key(&id) {
+        return Err(Error::Registered);
+    }
+    registry.insert(id, device);
+
+    Ok(())
+}
+
+/// Unregisters an existing block device, returning the device if found.
+pub fn unregister(id: DeviceId) -> Result<Arc<dyn BlockDevice>, Error> {
+    DEVICE_REGISTRY
         .lock()
-        .get(str)
-        .cloned()
+        .remove(&id.as_encoded_u64())
+        .ok_or(Error::NotFound)
+}
+
+/// Collects all block devices.
+pub fn collect_all() -> Vec<Arc<dyn BlockDevice>> {
+    DEVICE_REGISTRY.lock().values().cloned().collect()
 }
 
-pub fn all_devices() -> Vec<(String, Arc<dyn BlockDevice>)> {
-    let block_devs = COMPONENT.get().unwrap().block_device_table.lock();
-    block_devs
-        .iter()
-        .map(|(name, device)| (name.clone(), device.clone()))
-        .collect()
+/// Looks up a block device of a given device ID.
+pub fn lookup(id: DeviceId) -> Option<Arc<dyn BlockDevice>> {
+    DEVICE_REGISTRY
+        .lock()
+        .values()
+        .find(|dev| dev.id() == id)
```

Comment:
Use the more efficient lookup, `DEVICE_REGISTRY.lock().get(&dev.id())`, rather than iterating all devices.